lib/take_elevated.lua

```diff
@@ -0,0 +1,98 @@
+local tokens_per_ms               = tonumber(ARGV[1])
+local bucket_size                 = tonumber(ARGV[2])
+local tokens_to_take              = tonumber(ARGV[3])
+local ttl                         = tonumber(ARGV[4])
+local drip_interval               = tonumber(ARGV[5])
+local erl_tokens_per_ms           = tonumber(ARGV[6])
+local erl_bucket_size             = tonumber(ARGV[7])
+local erl_activation_period_seconds  = tonumber(ARGV[8])
+
+-- the key to use for pulling last bucket state from redis
+local lastBucketStateKey = KEYS[1]
+
+-- the key for checking in redis if elevated rate limits (erl) were activated earlier
+local erlKey = KEYS[2]
+local is_erl_activated = redis.call('EXISTS', erlKey)
+
+-- get current bucket state
+local current = redis.pcall('HMGET', lastBucketStateKey, 'd', 'r')
+if current.err ~= nil then
+    current = {}
+end
+
+-- get current time from redis, to be used in new bucket size calculations later
+local current_time = redis.call('TIME')
+local current_timestamp_ms = current_time[1] * 1000 + current_time[2] / 1000
+
+local function calculateNewBucketContent(current, tokens_per_ms, bucket_size, current_timestamp_ms)
+    if current[1] and tokens_per_ms then
+        -- drip bucket
+        local last_drip = current[1]
+        local content = current[2]
+        local delta_ms = math.max(current_timestamp_ms - last_drip, 0)
+        local drip_amount = delta_ms * tokens_per_ms
+        return math.min(content + drip_amount, bucket_size)
+    elseif current[1] and tokens_per_ms == 0 then
+        -- fixed bucket
+        return current[2]
+    else
+        -- first take of the bucket
+        return bucket_size
+    end
+end
+
+-- Enable verbatim replication to ensure redis sends script's source code to all masters
+-- managing the sharded database in a clustered deployment.
+-- https://redis.io/docs/interact/programmability/eval-intro/#:~:text=scripts%20debugger.-,Script%20replication,-In%20standalone%20deployments
+redis.replicate_commands()
```

Comment:
Is it still required in newer Redis versions?

Reply:
Yes. We're using Redis 6. In Redis 5.0, effects replication became the default mode - but we're interested in the verbatim mode here.